### PR TITLE
[Serve] Refactor `ReplicaQueueLengthAutoscalingPolicy` into `AutoscalingPolicyManager` and policy function

### PR DESCRIPTION
**Pull request by GeneDer**
## Why are these changes needed?

Refactor the existing `ReplicaQueueLengthAutoscalingPolicy` into `AutoscalingPolicyManager` to managing the lifecycle of policy call and applies bounds. `AutoscalingPolicyManager` also served as interface layer between the policy and the `DeploymentState`. Refactored the rest of core policy logics into `replica_queue_length_autoscaling_policy` policy function for use by `AutoscalingPolicyManager`. 

## Related issue number

Second PR for https://github.com/ray-project/ray/issues/41135

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/serve/_private/autoscaling_policy.py [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+            target_capacity,
+        )
+        decision_num_replicas = self.policy(
+            curr_target_num_replicas=curr_target_num_replicas,
+            current_num_ongoing_requests=current_num_ongoing_requests,
+            current_handle_queued_queries=current_handle_queued_queries,
+            config=self.config,
+            capacity_adjusted_min_replicas=capacity_adjusted_min_replicas,
+            capacity_adjusted_max_replicas=capacity_adjusted_max_replicas,
+            policy_states=self.policy_states,
```
Comment (edoakes) [inline python/ray/serve/_private/autoscaling_policy.py:60]: nit: `policy_state` (not plural)

**Review by edoakes — APPROVED** [review]

File: python/ray/serve/autoscaling_policy.py [hunk 2]
```diff
@@ -92,47 +89,15 @@ def _calculate_desired_num_replicas(
...
-@PublicAPI(stability="alpha")
-class ReplicaQueueLengthAutoscalingPolicy(AutoscalingPolicy):
+def replica_queue_length_autoscaling_policy(
+    curr_target_num_replicas: int,
+    current_num_ongoing_requests: List[float],
+    current_handle_queued_queries: float,
+    config: Optional[AutoscalingConfig],
+    capacity_adjusted_min_replicas: int,
+    capacity_adjusted_max_replicas: int,
+    policy_state: Dict[str, Any],
```
Comment (alexeykudinkin) [inline python/ray/serve/autoscaling_policy.py:99]: This seems quite verbose and not very flexible for us to add new sources.

Let's create a data class incorporating all of that (something along the lines of `AutoscalingPolicyInput`)

File: python/ray/serve/autoscaling_policy.py [hunk 3]
```diff
@@ -142,136 +107,59 @@ class ReplicaQueueLengthAutoscalingPolicy(AutoscalingPolicy):
     `get_decision_num_replicas` is called once every CONTROL_LOOP_PERIOD_S
     seconds.
```
Comment (alexeykudinkin) [inline python/ray/serve/autoscaling_policy.py:108]: Why do we need such assumption?

File: python/ray/serve/autoscaling_policy.py [hunk 4]
```diff
@@ -142,136 +107,59 @@ class ReplicaQueueLengthAutoscalingPolicy(AutoscalingPolicy):
...
+
+        # Only actually scale the replicas if we've made this decision for
+        # 'scale_down_consecutive_periods' in a row.
+        if decision_counter < -int(config.downscale_delay_s / CONTROL_LOOP_PERIOD_S):
+            decision_counter = 0
+            decision_num_replicas = desired_num_replicas
+
+    # Do nothing.
+    else:
+        decision_counter = 0
```
Comment (alexeykudinkin) [inline python/ray/serve/autoscaling_policy.py:159]: This should be deferred to PolicyManager to whether accept, reject or delay scaling recommendation produced by the policy. Policy by itself should be stateless, though could be accepting historical data (say last 5min of rersource usage) to take its decision.

File: python/ray/serve/_private/autoscaling_policy.py [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+        if self.config:
+            self.policy = self.config.get_policy()
+
+    def should_autoscale(self) -> bool:
+        """Returns whether autoscaling should be performed."""
+        return self.policy is not None
+
+    def get_decision_num_replicas(
+        self,
+        curr_target_num_replicas: int,
```
[the remark this answers is not in the document]
Comment (GeneDer) [reply python/ray/serve/_private/autoscaling_policy.py:32]: 👍 will rename in the upcoming PR

File: python/ray/serve/_private/autoscaling_policy.py [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+        self.policy = None
+        self.policy_state = {}
+        self._create_policy()
+
+    def _create_policy(self):
+        """Creates an autoscaling policy based on the given config."""
+        if self.config:
+            self.policy = self.config.get_policy()
+
+    def should_autoscale(self) -> bool:
```
[the remark this answers is not in the document]
Comment (GeneDer) [reply python/ray/serve/_private/autoscaling_policy.py:26]: This is refactored from https://github.com/ray-project/ray/pull/42242/files#diff-c4c2583a4c2f3a3c87ada6faebd0b2dfef404e165df133a11195be5d65fcb387L1267 and keeping the naming. I feel we can change both places to `is_autoscaling_policy_enabled` to be more descriptive. What do you think?

File: python/ray/serve/autoscaling_policy.py (hunk 2 above)
Comment (GeneDer) [reply python/ray/serve/autoscaling_policy.py:99]: Yes, the upcoming PR will add an `AutoscalingContext` class to wrap all the inputs :)

File: python/ray/serve/autoscaling_policy.py (hunk 3 above)
Comment (GeneDer) [reply python/ray/serve/autoscaling_policy.py:108]: The logic for our default autoscaling policy has not changed and I think the comment is still suitable. Basically it's assuming controller's event loop runs around 0.1 seconds and use that to delay upscaling/ downscaling. I think we can follow up to clean up some of those logics later, but for the purpose of adding the custom autoscaling functionality I rather not touch this right now.

File: python/ray/serve/autoscaling_policy.py (hunk 4 above)
Comment (GeneDer) [reply python/ray/serve/autoscaling_policy.py:159]: Totally agreed the policy itself should be stateless and that's why it's been refactored to just a function. However, I don't think the policy manager should decide how fast or slow it the replica scales. There might be usecase that the user is tracking some sort of session count on the main page and wanting to warm up/ cool down the ML service by scaling the deployment replica up/ down accordingly. I don't think we should stop customer from quickly scale up/ down by moving this logics out.

Comment (alexeykudinkin) [reply python/ray/serve/autoscaling_policy.py:159]: Yes, policy should be able to control how fast we auto-scale, but policy shouldn't be relying on the frequency of its invocation (`CONTROL_LOOP_PERIOD_S`)

Comment (GeneDer) [reply python/ray/serve/autoscaling_policy.py:159]: Agreed, this is currently out of scope of the custom autoscaling project, but we can follow up to rewrite this logics to use the exact time when autoscale last happened :)

Comment (alexeykudinkin) [reply python/ray/serve/autoscaling_policy.py:159]: We can persist history w/in policy_state for ex
https://github.com/ray-project/ray/pull/42284#discussion_r1449405138

**Pull request merged**